### PR TITLE
Implement streaming support from bitbucket.

### PR DESCRIPTION
**Pull request by atl-dyon**
This PR seeks to introduce support for the plugin to retrieve files from bitbucket using the streaming endpoint provided to us. The caller is then presented with an InputStream that they may use. 
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketFilePathClient.java [hunk 1]
```diff
@@ -42,5 +44,19 @@ public interface BitbucketFilePathClient {
...
+     * @param scmFile the file to retrieve
+     * @return the bytes of the file in an {@link InputStream} object.
+     * @throws AuthorizationException     if the credentials did not allow access to the given url
+     * @throws NoContentException         if the server did not respond with a body
+     * @throws ConnectionFailureException if the server did not respond
+     * @throws NotFoundException          if the url does not exist, or there is no file at the requested url
+     * @throws BadRequestException        if the request was malformed and thus rejected by the server
+     * @throws ServerErrorException       if the server failed to process the request
+     * @throws BitbucketClientException   for all errors not already captured
+     */
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketFilePathClient.java:60]: `@since` it helps us so for interfaces we tend to add them.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketFilePathClient.java [hunk 2]
```diff
@@ -42,5 +44,19 @@ public interface BitbucketFilePathClient {
      * @throws ServerErrorException       if the server failed to process the request
      * @throws BitbucketClientException   for all errors not already captured
      */
-    String getFileContent(BitbucketSCMFile scmFile);
+    String getFileContent(BitbucketSCMFile scmFile) throws IOException;
+
+    /**
+     * Retrieve the bytes of a file in a repository. The bytes are encapsulated in an {@link InputStream} object.
+     * @param scmFile the file to retrieve
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketFilePathClient.java:51]: newline please.

Also should probably add something about caller being responbile for closing the stream, just to make it clear for future selves.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRequestExecutor.java [hunk 3]
```diff
@@ -181,6 +192,14 @@ private <T> BitbucketResponse<T> makeGetRequest(HttpUrl url, ObjectReader<T> rea
                 }, addCredentials(additionalConfig));
     }
 
+    private InputStream streamGetRequest(HttpUrl url, RequestConfiguration... additionalConfig) {
+        return new ByteArrayInputStream(httpRequestExecutor.executeGet(url,
+                response -> {
+                    ensureNonEmptyBody(response);
+                    return response.body().bytes();
+                }, addCredentials(additionalConfig)));
+    }
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRequestExecutor.java:201]: Not keen on wrapping the response into a `ByteArrayInputStream` if we can avoid it. Can we not "just" do

```java
 return httpRequestExecutor.executeGet(url,
                response -> {
                    ensureNonEmptyBody(response);
                    return response.body().byteStream();
                }, addCredentials(additionalConfig));
    }
    
```
?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/HttpRequestExecutor.java [hunk 4]
```diff
@@ -87,6 +89,6 @@ <T> T executePut(HttpUrl url, String requestBodyAsJson, ResponseConsumer<T> cons
 
         ResponseConsumer<Void> EMPTY_RESPONSE = response -> null;
 
-        T consume(Response response);
+        T consume(Response response) throws IOException;
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/HttpRequestExecutor.java:92]: if we go with my suggestion above, I think we can revert this change.

**Review by dkjellin — CHANGES_REQUESTED** [review]
Can we also remove some of the code we have for paging through the file contents with this change?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketFilePathClient.java [hunk 5]
```diff
@@ -42,5 +44,19 @@ public interface BitbucketFilePathClient {
      * @throws ServerErrorException       if the server failed to process the request
      * @throws BitbucketClientException   for all errors not already captured
      */
-    String getFileContent(BitbucketSCMFile scmFile);
+    String getFileContent(BitbucketSCMFile scmFile) throws IOException;
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketFilePathClient.java:47]: If we go with the suggested changes, I don't think we need throws `IOException` here?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketFilePathClientImpl.java [hunk 6]
```diff
@@ -49,7 +54,7 @@ public List<SCMFile> getDirectoryContent(BitbucketSCMFile directory) {
     }
 
     @Override
-    public String getFileContent(BitbucketSCMFile file) {
+    public String getFileContent(BitbucketSCMFile file) throws IOException {
         HttpUrl url = getUrl(file);
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketFilePathClientImpl.java:58]: I think should use the streaming endpoint and gather the contents up in a `String` at the end, but I think we don't even need it any more, so I lean more towards just removing this along with the `FileNextPageFetcher` and `BitbucketFilePage` objects

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRequestExecutor.java (hunk 3 above)
Comment (atl-dyon) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRequestExecutor.java:201]: I tried playing around with that, however the stream is closed once we leave the lambda function. This way we are essentially creating a copy of the stream that is usable by the caller. It's not ideal but its the best i could come up with.

**Comment by dkjellin** [comment]
> Fixes [JENKINS-66829](https://issues.jenkins.io/browse/JENKINS-66829), I think.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/HttpRequestExecutor.java [hunk 7]
```diff
@@ -45,6 +45,25 @@ public interface HttpRequestExecutor {
...
+     * @param <T>              result that consumer wish to return
+     * @return result
+     * @throws AuthorizationException     if the credentials did not allow access to the given url
+     * @throws ConnectionFailureException if the server did not respond
+     * @throws NotFoundException          if the requested url does not exist
+     * @throws BadRequestException        if the request was malformed and thus rejected by the server
+     * @throws ServerErrorException       if the server failed to process the request
+     * @throws BitbucketClientException   for all errors not already captured
+     *
+     * @since 4.43
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/HttpRequestExecutor.java:63]: `3.3.3` I believe?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/HttpRequestExecutor.java [hunk 8]
```diff
@@ -45,6 +45,25 @@ public interface HttpRequestExecutor {
...
+     */
+    <T> T executeStreamingGet(HttpUrl url, ResponseConsumer<T> consumer, RequestConfiguration... additionalConfig);
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/HttpRequestExecutor.java:65]: Hmm.. wouldn't this _always_ return an `InputStream`? How could I be streamlining and not return a stream? I think this signature needs to change to 
`InputStream executeStreamingGet(HttpUrl url,  RequestConfiguration... additionalConfig);`

The implementation throws away any knowledge of `T` and also throws away the `consumer`  without using it.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/http/HttpRequestExecutorImpl.java [hunk 9]
```diff
@@ -113,38 +155,27 @@ private void handleError(int responseCode, @Nullable String body, Headers header
...
-                }
+            response = handleRequest(request);
+        } finally {
+            if (response != null && response.body() != null) {
+                response.close();
             }
-            throw e;
         }
-        throw new UnhandledErrorException("Unhandled error", -1, null);
+        return consumer.consume(response);
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/http/HttpRequestExecutorImpl.java:166]: this needs to move _inside_ the try/catch block. You close the response before the body may have been fully consumed. It probably works really well in tests because bodies are small and fast, in real life I imagine this could cause issues.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/http/HttpRequestExecutorImpl.java [hunk 10]
```diff
@@ -74,6 +81,41 @@ private <T> T executeRequest(Request.Builder requestBuilder, ResponseConsumer<T>
         return performRequest(requestBuilder.build(), consumer);
     }
 
+    private Response handleRequest(Request request) {
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/http/HttpRequestExecutorImpl.java:84]: `makeRequest`?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/http/HttpRequestExecutorImpl.java [hunk 11]
```diff
@@ -113,38 +155,27 @@ private void handleError(int responseCode, @Nullable String body, Headers header
...
+    }
+
+    private InputStream performStreamingGet(HttpUrl url, RequestConfiguration... additionalConfigs) {
+        Set<RequestConfiguration> additionalConfig = toSet(additionalConfigs);
+        Request.Builder requestBuilder = new Request.Builder().url(url);
+        additionalConfig.forEach(config -> config.apply(requestBuilder));
+        return performStreamingRequest(requestBuilder.build());
+    }
+
+    private InputStream performStreamingRequest(Request request) {
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/http/HttpRequestExecutorImpl.java:176]: can I propose we just inline this method? It is called just once from the method above and it is tiny. We can extract things out when we need it

**Review by dkjellin — CHANGES_REQUESTED** [review]

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/HttpRequestExecutor.java [hunk 12]
```diff
@@ -45,6 +47,24 @@ public interface HttpRequestExecutor {
      */
     <T> T executeGet(HttpUrl url, ResponseConsumer<T> consumer, RequestConfiguration... additionalConfig);
 
+    /**
+     * Executes a streaming get call to a given URL.
+     *
+     * @param url              The URL to hit on bitbucket server end
+     * @param additionalConfig additional configuration, such as credentials, that will be used in making calls
+     * @param consumer         on successful execution, {@link Response} will be passed to consumer
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/HttpRequestExecutor.java:55]: `consumer` is unused, and should be removed I think. It doesn't quite make sense to both return an `InputStream` AND take in a `consumer`

**Review by dkjellin — APPROVED** [review]
I'm going to hit approve, as I think this is there, it is just the unused `consumer` left, but happy for us to fix that as a separate PR if we want to, it is a trivial change.

**Review by mhenschke-atl — APPROVED** [review]
